cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -2425,43 +2414,69 @@ x"
         let e = parse(
             YaccKind::Grmtools,
             "%start A
+                %start A
+                %start A
              %%
              A -> (): 'a1';
-             B -> (): 'bb';
-             A -> (): 'a2';
-             C -> (): 'cc';
-             A -> (): 'a3';
             ",
         )
         .expect_err("parse successful while expecting error");
         assert_eq!(
             e,
             vec![YaccGrammarError {
-                kind: YaccGrammarErrorKind::DuplicateRule,
-                spans: vec![Span::new(38, 39), Span::new(94, 95), Span::new(150, 151)],
+                kind: YaccGrammarErrorKind::DuplicateStartDeclaration,
+                spans: vec![Span::new(7, 8), Span::new(32, 33), Span::new(57, 58)],
             }]
         );
     }
 
     #[test]
-    fn test_duplicate_rules_and_missing_arrow() {
-        let src = "
-        %%
-        A -> () : 'a1';
-        A -> () : 'a2';
-        B -> () : 'b1';
-        A -> () : 'a3';
-        B -> () : 'b2';
-        B";
+    fn test_duplicate_rule() -> Result<(), Vec<YaccGrammarError>> {
+        let ast = parse(
+            YaccKind::Grmtools,
+            "%token A B D
+%%
+Expr -> () : %empty | A;
+Expr -> () : B | 'C';
+Expr -> () : D;
+",
+        )?;
+        let expr_rule = ast.get_rule("Expr").unwrap();
+        let mut prod_names = HashSet::new();
+        for pidx in &expr_rule.pidxs {
+            for sym in &ast.prods[*pidx].symbols {
+                let name = match sym {
+                    Symbol::Token(name, _) | Symbol::Rule(name, _) => name.clone(),
+                };
+                prod_names.insert(name);
+            }
+        }
+        let mut expected_prod_names = HashSet::new();
```

Comment:
Could this be something like `assert_eq!(prod_names, HashSet::from(["A", ..., "D"]));` ? [I don't know if `HashSet` supports such a constructor, but you never know! I suppose `collect` works at a push?]

Reply:
Yeah it implements `FromIterator`, so yeah collect should work!